spec/acceptance/cloudkeeper/aws/grpc_api_mock_spec.rb

```diff
@@ -24,59 +27,107 @@ module Aws
-              when 1
-                { status_code: 200, headers: {}, body: '' }
-              end
-            end)
-            core_connector.add_appliance(appliance, nil)
+            before do
+              allow(::Aws::S3::Object).to receive(:new).and_return(s2client_mock)
+              allow(s2client_mock).to receive(:upload_file) { |file_name| @file_name = file_name }
+              allow(s2client_mock).to receive(:exists?).and_return(false)
+              core_connector.add_appliance(appliance, nil)
+            end
+
+            it 'uploads image' do
+              expect(@file_name).to eq(image_location)
+            end
           end
 
-          it 'uploads image' do
-            expect(@body).to eq(stub_data)
+          context 'with not existing file' do
+            let(:image_location) { File.join(SPEC_DIR, 'notexist') }
+
+            it 'raises error' do
+              expect { core_connector.add_appliance(appliance, nil) }.to \
+                raise_error(GRPC::BadStatus)
+            end
           end
-        end
 
-        context 'with invalid image uri' do
-          let(:image_uri) { 'https://wrongserver.com/test.img' }
+          context 'with already uploaded image' do
+            let(:image_location) { File.join(SPEC_DIR, 'mock_image') }
+
+            before do
+              s3.client.stub_responses(:head_object,
+                                       status_code: 200,
+                                       headers: {},
+                                       body: stub_data)
+            end
 
-          it 'does not upload image' do
-            expect { core_connector.add_appliance(appliance, nil) }.to \
-              raise_error(GRPC::BadStatus)
+            it 'does not upload image' do
+              expect { core_connector.add_appliance(appliance, nil) }.to \
+                raise_error(GRPC::BadStatus)
+            end
           end
         end
 
-        context 'with already uploaded image' do
-          let(:image_uri) { 'https://testserver.com/test.img' }
+        context 'when in remote mode' do
+          let(:image_mode) { :REMOTE }
 
           before do
-            s3.client.stub_responses(:head_object,
-                                     status_code: 200,
-                                     headers: {},
-                                     body: stub_data)
+            # Stubs for downloading images
+            stub_request(:get, /testserver.com/).to_return(body: stub_data)
+            stub_request(:get, /wrongserver.com/).to_return(status: 404)
           end
 
-          it 'does not upload image' do
-            expect { core_connector.add_appliance(appliance, nil) }.to \
-              raise_error(GRPC::BadStatus)
+          context 'with valid appliance' do
+            let(:image_location) { 'https://testserver.com/test.img' }
+
+            before do
+              count = 0
+              s3.client.stub_responses(:head_object, lambda do |_|
```

Comment:
Why do you need this monstrosity? And why are you calling `core_connector.add_appliance(appliance, nil)` in the `before` block?

Reply:
The "monstrosity" stubs AWS responses to `obj.exists?`, so that Cloudkeeper-AWS thinks that file does not exist and then that file does exist.

`core_connector.add_appliance` is called in before block so that this context can have multiple `it` blocks in the future. For example now it checks if correct image was uploaded, but maybe I'll want check how many requests were made.